Fix texture scroll, optimise texture matrix operations a bit

VReaperV opened the pull request:
Fixes #1278.

The previous code assumed that a texture with a scroll textureMod would always wrap around when `scroll * backEnd.refdef.floatTime == floor( scroll * backEnd.refdef.floatTime )`. This is incorrect if the texture matrix had already been modified by another texMod, so the relevant values are no longer 1.0f and 0.0f. Fix this by using `scrollPeriod[0] / ( matrix[0] + matrix[4] )` and `scrollPeriod[0] / ( matrix[0] + matrix[4] )` to get the actual time at which wrap-around for `x` and `y` will occur, and use `fmodf( backEnd.refdef.floatTime, period )` to get the part that is equivalent to just using `backEnd.refdef.floatTime`, but without the value getting continuously larger.

Also NUKED some useless matrix computations from texMod scroll, and added some functions to only calculate the relevant part of a matrix for texture matrices.

This change can be observed on:
Map `tds`, `r_forceRendererTime 49999` and `r_forceRendererTime 50000`:
Old:
https://imgsli.com/MzQ0MjYy
New:
https://imgsli.com/MzQ0MjY1
Map `pulse`, `r_forceRendererTime 50000` and `r_forceRendererTime 50001`:
Old:
https://imgsli.com/MzQ0MjY0
New:
https://imgsli.com/MzQ0MjY2

slipher commented:
At another spot on Pulse, with this PR I get a new instance of the scrolling glitch where there wasn't one before. The lorem ipsum star wars text thing at the spectator spawn only shows the first 2 lines then resets.

VReaperV commented:
Oh, right, there were some divs by 0 and a wrong index. Fixed now.

slipher reviewed (commented):
I tried combining a rotate with a scroll and it still suffers from discontinuity:

```
textures/pulse/title
{
	qer_editorimage textures/pulse/title.jpg

	noPicMip

	surfaceparm noimpact
	surfaceparm nolightmap
   	surfaceparm trans
	surfaceparm nomarks
	{
		map textures/pulse/title.jpg
		tcMod rotate 5
		tcMod scroll 0 0.025
		blendFunc add
	}
}
```

This could still be merged as-is though, as it fixes some things and doesn't break anything that is not already broken AFAIK.

VReaperV commented on src/engine/qcommon/q_shared.h:
```diff
@@ -877,6 +877,38 @@ inline vec_t VectorNormalize2( const vec3_t v, vec3_t out )
 		MatrixFromAngles( m, angles[ PITCH ], angles[ YAW ], angles[ ROLL ] );
 	}
 
+inline void Mat3x2MultiplyScale( matrix_t m, const float x, const float y ) {
```
I've moved these to `tr_shade_calc.cpp` now.

VReaperV commented:
> I tried combining a rotate with a scroll and it still suffers from discontinuity:
> 
> ```
> textures/pulse/title
> {
> 	qer_editorimage textures/pulse/title.jpg
> 
> 	noPicMip
> 
> 	surfaceparm noimpact
> 	surfaceparm nolightmap
>    	surfaceparm trans
> 	surfaceparm nomarks
> 	{
> 		map textures/pulse/title.jpg
> 		tcMod rotate 5
> 		tcMod scroll 0 0.025
> 		blendFunc add
> 	}
> }
> ```
> 
> This could still be merged as-is though, as it fixes some things and doesn't break anything that is not already broken AFAIK.

Fixed, also fixed some more potential divisions by 0.

slipher commented:
The rotation+scroll shader I wrote above got more buggy with the latest code. It still makes discontinuous jumps, and also seems to accelerate faster and faster over time.

VReaperV commented:
Hmm, I probably messed something up before pushing to the remote.

VReaperV commented:
> The rotation+scroll shader I wrote above got more buggy with the latest code. It still makes discontinuous jumps, and also seems to accelerate faster and faster over time.

Fixed now.

slipher commented:
I got a build error as does the CI

```
/usr/bin/g++ -DARCH_STRING=amd64 -DBUILD_ENGINE -DBUILD_GRAPHICAL_CLIENT -DDAEMON_ARCH_amd64 -DDAEMON_BUILD_Release -DDAEMON_CXX_COMPILER_GCC=1 -DDAEMON_CXX_COMPILER_STRING=GCC_9.4.0:g++ -DDAEMON_C_COMPILER_GCC=1 -DDAEMON_C_COMPILER_STRING=GCC_9.4.0:gcc -DDAEMON_OpenGL_ABI=LEGACY -DDAEMON_OpenGL_ABI_LEGACY -DDAEMON_USE_ARCH_INTRINSICS=1 -DDAEMON_USE_ARCH_INTRINSICS_amd64=1 -DDAEMON_USE_ARCH_INTRINSICS_i686=1 -DDAEMON_USE_COMPILER_CUSTOMIZATION=1 -DDAEMON_USE_COMPILER_INTRINSICS=1 -DNACL_ANDROID=0 -DNACL_ARCH_STRING=amd64 -DNACL_BUILD_ARCH=x86 -DNACL_BUILD_SUBARCH=64 -DNACL_FREEBSD=0 -DNACL_LINUX=1 -DNACL_OSX=0 -DNACL_WINDOWS=0 -DTHIS_IS_NOT_A_DEBUG_BUILD -DUSE_CURSES -DUSE_CURSES_NCURSES -DUSE_MUMBLE -DUSE_SMP -I/home/vsts/work/1/s/src -I/home/vsts/work/1/s/libs -I/home/vsts/work/1/s/libs/nacl -I/usr/include/ncursesw -I/usr/include/opus -I/usr/include/freetype2 -I/usr/include/AL -I/home/vsts/work/1/s/src/engine -I/home/vsts/work/1/s/build/embed_data -isystem /usr/include/SDL2 -std=gnu++14 -ffast-math -fvisibility=hidden -fno-strict-aliasing -fno-gnu-unique -pthread -march=x86-64 -mtune=generic -mcx16 -O3 -DNDEBUG -D_FORTIFY_SOURCE=2 -Wall -Wextra -Werror -Wshadow=local -Wno-pragmas -Wno-unknown-pragmas -Wno-error=old-style-cast -Woverloaded-virtual -Wstrict-null-sentinel -Wno-error=sign-compare -MD -MT CMakeFiles/client-objects.dir/src/engine/renderer/tr_shade_calc.cpp.o -MF CMakeFiles/client-objects.dir/src/engine/renderer/tr_shade_calc.cpp.o.d -o CMakeFiles/client-objects.dir/src/engine/renderer/tr_shade_calc.cpp.o -c /home/vsts/work/1/s/src/engine/renderer/tr_shade_calc.cpp
/home/vsts/work/1/s/src/engine/renderer/tr_shade_calc.cpp: In function ‘void RB_CalcTexMatrix(const textureBundle_t*, vec_t*)’:
/home/vsts/work/1/s/src/engine/renderer/tr_shade_calc.cpp:787:41: error: taking address of temporary array
  787 |      ComputeTextureWrapModifer( matrix, vec2_t{ x, y }, modifier );
      |                                         ^~~~~~~~~~~~~~
/home/vsts/work/1/s/src/engine/renderer/tr_shade_calc.cpp:849:41: error: taking address of temporary array
  849 |      ComputeTextureWrapModifer( matrix, vec2_t { x, y }, modifier );
      |                                         ^~~~~~~~~~~~~~~
```

VReaperV commented:
Oh, right, fixed.

slipher commented:
With the shader I posted above, I still get the effect of faster scrolling over time. Like use `timescale 10` to skip ahead in time a bit to 10:00 game time, then compare how fast it scrolls with how fast it scrolled at time 0.

VReaperV commented:
> With the shader I posted above, I still get the effect of faster scrolling over time. Like use `timescale 10` to skip ahead in time a bit to 10:00 game time, then compare how fast it scrolls with how fast it scrolled at time 0.

So this is actually bugged on master. When removing the `x = x - floor( x );` .. part, this speeding-up effect is also present on master.

It can be observed when using `r_forceRendererTime` too.

E. g. with `r_forceRendererTime 2030000` and `r_forceRendererTime 2030010`:
Master:
![unvanquished_2025-02-23_032800_000](https://github.com/user-attachments/assets/4d97ca63-c59d-43b1-bbe8-c289b1a5d9ff)
![unvanquished_2025-02-23_032805_000](https://github.com/user-attachments/assets/041d779d-0565-48c0-acc6-cd80a169b5a4)

Master without the `floor()` parts:
![unvanquished_2025-02-23_032713_000](https://github.com/user-attachments/assets/014c25fd-3bfb-4e54-9692-01dfa72161d5)
![unvanquished_2025-02-23_032800_000](https://github.com/user-attachments/assets/153f77ab-dc0b-44b0-880a-acf96b57a422)

This branch:
![unvanquished_2025-02-23_032855_000](https://github.com/user-attachments/assets/b94fbe3b-3151-4099-a6a0-baacccac866e)
![unvanquished_2025-02-23_032859_000](https://github.com/user-attachments/assets/5e912f95-3cf8-446c-b446-a8481edc6f09)

VReaperV commented:
Well, it doesn't *exactly* line up, which might be due to float precision issues. But it definitely seems to move the correct amount/with the correct speed.

slipher commented:
I suppose there is a lot of ambiguity as to what these operations mean when combined. We could interpret scrolling as (a) translating on an infinite plane tiled with the texture or (b) wrapping a fixed-size rectangle. Option (a) would explain the speeding up since you get arbitrarily fast acceleration as you go farther from origin point of the rotation. Master does something perhaps more like (b) but there's probably no way to do it smoothly with a single texture matrix, explaining the discontinuity bugs.

LGTM since we don't know any affected shaders anyway.

VReaperV commented:
(rebased + removed the accidental new-line)

VReaperV commented:
> I suppose there is a lot of ambiguity as to what these operations mean when combined. We could interpret scrolling as (a) translating on an infinite plane tiled with the texture or (b) wrapping a fixed-size rectangle. Option (a) would explain the speeding up since you get arbitrarily fast acceleration as you go farther from origin point of the rotation. Master does something perhaps more like (b) but there's probably no way to do it smoothly with a single texture matrix, explaining the discontinuity bugs.
> 
> LGTM since we don't know any affected shaders anyway.

Yeah, I haven't seen any shader actually using something like this.
